src/instance.rs

```diff
@@ -39,6 +39,16 @@ pub unsafe trait PyNativeType: Sized {
     }
 }
 
+/// Trait to denote types which have a GIL-bound reference, e.g. PyAny has &'a PyAny.
+///
+/// # Safety
+/// `Py<T>` will hand out references to `Self::AsRefTarget`. This _must_ have the
+/// same layout as `*mut ffi::PyObject`.
+pub unsafe trait HasPyGilBoundRef {
```

Comment:
IMO the name could use some work. I don't have a name that is obviously good, but what about something like `AsGilBorrow`?

Reply:
Sure thing. I think we can sit on this and think for a bit. I'm a bit lost for a good name too. `StorableInPy`?

Reply:
Just add completely different direction for the bikeshed: `PyHeapType` meaning that can be stored on the Python heap (i.e. in `Py<T>`) and thereby shared references must be bound to the GIL (as it protects access to the heap).